### PR TITLE
Add formatter progress tracking to CI

### PR DESCRIPTION
**Pull request by konstin**
**Summary** Add a formatter progress testing script to CI. This script will 1) print the black compability on each run 2) catch regressions wrt to formatter stability, emitting invalid syntax and other kinds of bugs (e.g. #5917) before they land on main 3) have an additional layer of real world tests when implementing new nodes or other new formatter code.

This is currently a bash script, i'm not sure if we want to keep it that way, or switch to e.g. the regular ecosystem scripts. The output separation of `format_dev` could also use some polishing. We should also consider pinning commits so we don't get spurious regression when they change their code.

**Test Plan** The script extends CI.

### CODE REVIEW

File: scripts/formatter_progress.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# an ML project
+if [ ! -d "$dir/transformers" ]; then git clone https://github.com/huggingface/transformers "$dir/transformers"; fi
+# type annotations
+if [ ! -d "$dir/typeshed" ]; then git clone https://github.com/python/typeshed "$dir/typeshed"; fi
+# python 3.11, typing and 100% test coverage
+if [ ! -d "$dir/warehouse" ]; then git clone https://github.com/pypi/warehouse "$dir/warehouse"; fi
+# django project
+if [ ! -d "$dir/zulip" ]; then git clone https://github.com/zulip/zulip "$dir/zulip"; fi
+
+for i in "$dir"/*/; do echo "# $(basename $i) $(git -C "$i" rev-parse HEAD)"; done
```
Comment (zanieb) [inline scripts/formatter_progress.sh:22]:
```suggestion
for i in "$dir"/*/; do echo "# $(basename "$i") $(git -C "$i" rev-parse HEAD)"; done
```

File: scripts/formatter_progress.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
```
Comment (zanieb) [inline scripts/formatter_progress.sh:1]:
```suggestion
#!/usr/bin/env bash
```

**Comment by konstin** [comment]
I think the main question remaining for me is: Should we pin the repos to a specific commit?

pro: the score may otherwise change due to downstream changes unrelated to out ours
pro: correct PRs may break if they add changes that break our formatter
con: we won't see any new python syntax with an old commit pinned
con: we won't see any black updates with an old commit pinned 

Effectively, i want lockfiles and dependabot for this, but that doesn't exist, so we'll have to go with either option

**Comment by konstin** [comment]
the result is now posted as github step summary (https://github.com/astral-sh/ruff/actions/runs/5636816309?pr=5919#summary-15269270283):

image

i wish i could show that without switching to the CI tab and scrolling down, but it's still good to have this in CI either way

**Comment by MichaReiser** [comment]
> I think the main question remaining for me is: Should we pin the repos to a specific commit?


I'm leaning toward pinning:

* Updating is low effort. We can do this every week
* It is otherwise unclear if the similarity index improve because of my changes or because of changes in the tested repositories
* Having all formatter PRs blocked by a new stability issue would be unfortunate. It also reduces trust in the check because there's always the chance that your change didn't introduce the instability.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -323,14 +323,19 @@ jobs:
     name: "Check formatter stability"
     runs-on: ubuntu-latest
     needs: determine_changes
-    if: needs.determine_changes.outputs.formatter == 'true'
+    # if: needs.determine_changes.outputs.formatter == 'true'
```
Comment (MichaReiser) [inline .github/workflows/ci.yaml:326]: Before merge: Revert

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -323,14 +323,19 @@ jobs:
     name: "Check formatter stability"
```
Comment (MichaReiser) [inline .github/workflows/ci.yaml:323]: Nit: Rename, considering that it now checks more than just the progress.

File: scripts/formatter_progress.sh [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+# django project
+if [ ! -d "$dir/zulip" ]; then git clone https://github.com/zulip/zulip "$dir/zulip"; fi
+
+for i in "$dir"/*/; do echo "# $(basename "$i") $(git -C "$i" rev-parse HEAD)"; done
+# build 5800521541e5e749d4429617420d1ef8cdb40b46
+# django 0016a4299569a8f09ff24053ff2b8224f7fa4113
+# transformers 5bb4430edc7df9f9950d412d98bbe505cc4d328b
+# typeshed 57c435cd7e964290005d0df0d9b5daf5bd2cbcb1
+# warehouse e72cca94e7ac0dbe095db5c2942ad9f2f51b30cc
+# zulip 1cd587d24be1d668fcf6d136172bfec69e35cb75
```
Comment (MichaReiser) [inline scripts/formatter_progress.sh:28]: What's this?

File: scripts/formatter_progress.sh [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+cargo run --bin ruff_dev -- format-dev --stability-check --error-file "$target/progress_projects_errors.txt" \
+  --multi-project "$dir" >"$target/progress_projects_report.txt"
+grep "similarity index" "$target/progress_projects_report.txt" | sort
```
Comment (MichaReiser) [inline scripts/formatter_progress.sh:32]: Nit: An alternative to grepping could have been to add a `--markdown` flag to the CLI that enables a markdown rendered summary

**Review by MichaReiser — APPROVED** [review]

**Pull request merged**